# 1. Heart Attack Prediction (60pt)

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 1. (2pt) Load the data and drop the variable slp, oldpeak and thall, since they do not have descriptions available. Do some basic sanity check (descriptive statistics; missing values; data types...). The data should contain 303 rows, and 11 columns.

In [2]:
heart = pd.read_csv("../data/heart.csv")
heart = heart.drop(["slp","oldpeak","thall"], axis=1)
heart.shape

(303, 11)

In [3]:
heart

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  caa  output
0     63    1   3     145   233    1        0       150     0    0       1
1     37    1   2     130   250    0        1       187     0    0       1
2     41    0   1     130   204    0        0       172     0    0       1
3     56    1   1     120   236    0        1       178     0    0       1
4     57    0   0     120   354    0        1       163     1    0       1
..   ...  ...  ..     ...   ...  ...      ...       ...   ...  ...     ...
298   57    0   0     140   241    0        1       123     1    0       0
299   45    1   3     110   264    0        1       132     0    0       0
300   68    1   0     144   193    1        1       141     0    2       0
301   57    1   0     130   131    0        1       115     1    1       0
302   57    0   1     130   236    0        0       174     0    1       0

[303 rows x 11 columns]

In [4]:
heart.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
caa         0
output      0
dtype: int64

In [5]:
heart.dtypes

age         int64
sex         int64
cp          int64
trtbps      int64
chol        int64
fbs         int64
restecg     int64
thalachh    int64
exng        int64
caa         int64
output      int64
dtype: object

In [6]:
heart['cp'] = pd.Categorical(heart.cp)
heart['caa'] = pd.Categorical(heart.caa)

## 2. (22pt) Construct a simple logistic regression with statsmodel.formula.api package. The outcome variable is output. The rest 10 variables (age, sex, cp, trtbps, chol, fbs, restecg, thalachh, exng, caa) in the dataset are predictors. Print out the marginal effect summary table and answer the following questions:

In [7]:
import statsmodels.formula.api as smf
m = smf.logit('output ~ age + sex + cp + trtbps + chol + fbs + restecg + thalachh + exng + caa', data=heart).fit()
m.summary()

Optimization terminated successfully.
         Current function value: 0.356701
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 output   No. Observations:                  303
Model:                          Logit   Df Residuals:                      287
Method:                           MLE   Df Model:                           15
Date:                Fri, 04 Mar 2022   Pseudo R-squ.:                  0.4824
Time:                        01:02:10   Log-Likelihood:                -108.08
converged:                       True   LL-Null:                       -208.82
Covariance Type:            nonrobust   LLR p-value:                 1.064e-34
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.4078      2.464      0.166      0.869      -4.421       5.237
cp[T.1]        1.5510      0.543      2.857      0.004       0.487       2.615
cp[T.2]        1.7048      0.441      3.865      0.000       0.840       2.569
cp[T.3]        1.8943      0.666      2.843      0.004       0.589       3.200
caa[T.1]      -1.8878      0.457     -4.134      0.000      -2.783      -0.993
caa[T.2]      -3.0263      0.645     -4.690      0.000      -4.291      -1.762
caa[T.3]      -2.4131      0.779     -3.098      0.002      -3.940      -0.887
caa[T.4]       0.4632      1.569      0.295      0.768      -2.613       3.539
age            0.0239      0.024      0.999      0.318      -0.023       0.071
sex           -2.1000      0.452     -4.648      0.000      -2.985      -1.214
trtbps        -0.0292      0.011     -2.755      0.006      -0.050      -0.008
chol          -0.0062      0.004     -1.565      0.118      -0.014       0.002
fbs            0.4568      0.502      0.910      0.363      -0.527       1.440
restecg        0.3251      0.340      0.955      0.340      -0.342       0.992
thalachh       0.0365      0.011      3.471      0.001       0.016       0.057
exng          -1.1710      0.418     -2.803      0.005      -1.990      -0.352
==============================================================================
"""

In [8]:
m.get_margeff().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                 output
Method:                          dydx
At:                           overall
==============================================================================
                dy/dx    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
cp[T.1]        0.1742      0.058      3.010      0.003       0.061       0.288
cp[T.2]        0.1915      0.045      4.270      0.000       0.104       0.279
cp[T.3]        0.2128      0.071      2.987      0.003       0.073       0.352
caa[T.1]      -0.2121      0.046     -4.648      0.000      -0.301      -0.123
caa[T.2]      -0.3400      0.062     -5.477      0.000      -0.462      -0.218
caa[T.3]      -0.2711      0.082     -3.322      0.001      -0.431      -0.111
caa[T.4]       0.0520      0.176      0.295      0.768      -0.294       0.398
age            0.0027      0.003      1.006      0.314      -0.003       0.008
sex           -0.2359      0.045     -5.291      0.000      -0.323      -0.149
trtbps        -0.0033      0.001     -2.891      0.004      -0.006      -0.001
chol          -0.0007      0.000     -1.582      0.114      -0.002       0.000
fbs            0.0513      0.056      0.914      0.361      -0.059       0.161
restecg        0.0365      0.038      0.959      0.338      -0.038       0.111
thalachh       0.0041      0.001      3.730      0.000       0.002       0.006
exng          -0.1315      0.045     -2.938      0.003      -0.219      -0.044
==============================================================================
"""

* (6pt) Interpret coefficient for sex. Is it statistically significant at significance level of 0.05?

Yes, it is statistically significant. $H0$ is outside the confidence interval, the p-value is less than the significance level of 0.05, and the z-value seems relatively high. All of these makes it statistically significant. Using **get_margeff()** to get the marginal effects, we get the following interpretation: Males are about 23.59% points less likely to get a heart attack, if other characteristics stays the same

* (6pt) Interpret coefficient for cp. Is it statistically significant at significance level of 0.05?

cp is also statistically significant because $H0$ does not fall inbetween the confidence interval, the p-value is less than the significant level, and the z-value is relatively high. Because cp is a categorical variable, each category has its own interpretation:
* cp[T.0]: Reference category
* cp[T.1]: People with atypical angina are 17.42% points more likely to get a heart attack than those with typical angina, given that all other coeffecients stays the same
* cp[T.2]: People with non-anginal pain are 19.15% points more likely to get a heart attack than those with typical angina, given that all other coefficients stays the same
* cp[T.3]: People that are asymptomatic are 21.28% points more likely to get a heart attack than those with typical angina, given that all other coefficients stays the same


* (6pt) Interpret the coefficient for age. Is it statistically significant at significance level of 0.05?

Age is not statistically significant because $H0$ is between the confidence interval, the p-value is above the significance level of 0.05, and the z-value is low. Age can be interpreted as the following: 1 year older means there is a .08% points decrease in the chance of getting a heart attack, given that all other characteristics stays equal.

* (4pt) What are the variables that are associated with lower chance of having a heart attack, and also are statistically significant? Do they intuitively make sense? Why?

The variables that are associated with having a lower chance of getting a heart attack are:
* sex - male/female
* trtbps - resting blood pressure
* chol - cholestoral
* exng - exercise induced angina
* caa - number of major vessels
These numbers do not make sense. From my knowledge, males are more likely to get a heart attack than females. Also, people with high cholestoral are at greater risk to get a heart attack.

## 3. (6pt) Now let’s construct the same model with sklearn package using the following code: LogisticRegression(penalty=’none’, solver=’newton-cg’).fit(X,y). Remember that sklearn package requires the predictor values (X) to be separated with the outcome variable (y). The predictor values also need to be in matrix format. Answer the following question:

In [9]:
x = heart.drop(["output"], axis=1)
y = heart["output"].values

In [10]:
from sklearn.linear_model import LogisticRegression
x_dummies = pd.get_dummies(x, drop_first=True, columns=["cp", "caa"])
m = LogisticRegression(penalty="none",solver="newton-cg").fit(x_dummies, y)

* (6pt) Print out the coefficient and intercept. Did you get the same intercept and coefficients as what you got from statsmodel.formula.api package? (you are supposed to!!)

In [11]:
print("Intercept: ", m.intercept_)
print("Coefficients: ", m.coef_)

Intercept:  [0.40774509]
Coefficients:  [[ 0.02385761 -2.09999102 -0.02920269 -0.00616402  0.4567869   0.3251173
   0.03647469 -1.17101514  1.5510426   1.70481698  1.8943255  -1.88776461
  -3.02631323 -2.41307618  0.46326514]]


Yes, the intercepts and coefficients are the same

## 4. (18pt) With the sklearn model, let’s do some predictions on the training set (the same X and y we used to train the model from Q3):

In [12]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(x,y, test_size=0.2)

In [13]:
logistic = LogisticRegression(penalty='none',solver='newton-cg').fit(train_X, train_y)
logistic.intercept_, logistic.coef_

(array([1.35438582]),
 array([[-0.02749316, -1.97838742,  0.9064679 , -0.02243834, -0.00677369,
          0.26884218,  0.73431038,  0.03937371, -0.92981693, -0.76184321]]))

In [14]:
test_y_pred_prob = logistic.predict_proba(test_X) #predict probability

* (6pt) The probability of having a heart attack: P (output=1|X). Print out the first 10 probabilities.

In [15]:
test_y_pred_prob[:10]

array([[0.22862983, 0.77137017],
       [0.08965974, 0.91034026],
       [0.06888269, 0.93111731],
       [0.33964775, 0.66035225],
       [0.03192148, 0.96807852],
       [0.774059  , 0.225941  ],
       [0.0537829 , 0.9462171 ],
       [0.32152002, 0.67847998],
       [0.68997557, 0.31002443],
       [0.48318524, 0.51681476]])

In [16]:
test_y_pred_prob_1 = test_y_pred_prob[:,1]
test_y_pred_prob_1[:10]

array([0.77137017, 0.91034026, 0.93111731, 0.66035225, 0.96807852,
       0.225941  , 0.9462171 , 0.67847998, 0.31002443, 0.51681476])

* (6pt) The outcome labels — that is, we directly predict whether or not the person will have a heart attack, instead of predicting the probability. Print out the first 10 labels.

In [17]:
test_y_pred_labels = logistic.predict(test_X)
test_y_pred_labels.shape

(61,)

* (6pt) Show the steps of how to convert from probabilities to the labels using threshold 0.5.

In [18]:
threshold = 0.5

## compare probabilities of y=1 with the threshold:
## - if probability > threshold, predict y=1
## - if probability <= threshold, predict y=0
1.0*(test_y_pred_prob_1>threshold)

array([1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 0.,
       1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0.,
       1., 0., 1., 1., 0., 1., 1., 1., 0., 1., 0., 0., 0., 1., 0., 1., 0.,
       0., 0., 0., 1., 0., 1., 1., 1., 1., 0.])

## 5. (6pt) Calculate the accuracy of the predicted output labels, compared with the true output labels. You can calculate it using your own code or using sklearn.metrics package. How to interpret the accuracy? Do you think the accuracy is high enough, such that you are comfortable deploying this model in real world to predict heart attack? (Hint: you should get about 80% accuracy.)

In [19]:
np.mean(test_y_pred_labels == test_y)*100

77.04918032786885

The accuracy means the percentage of predictions that are correct. In this case, it means that around 80% of our predictions perfectly predicted the outcome. Because this scenario involves someone's life, I would like the accuracy to be higher, maybe around 90%. 

## 6. (6pt) Create a confusion matrix on the training data. Calculate accuracy, precision, and recall based on the confusion matrix.

In [20]:
from sklearn.metrics import confusion_matrix

m1 = LogisticRegression().fit(x, y)
yhat = m1.predict(x)

cm = confusion_matrix(y, yhat)
cm

/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([[103,  35],
       [ 22, 143]])

In [21]:
np.mean(y == yhat) #accuracy

0.8118811881188119

In [22]:
from sklearn.metrics import precision_score
precision_score(y, yhat) #precision

0.8033707865168539

In [23]:
from sklearn.metrics import recall_score
recall_score(y, yhat) #recall

0.8666666666666667

# 2. Predict AirBnB Price (40pt)
## 1. (5pt) Replicate the model from PS06 question 2.7. Copy paste of your old code is OK.

In [24]:
airbnb = pd.read_csv('../data/airbnb-beijing-listings.csv.bz2', usecols=['price', 'bedrooms', 'accommodates', 'bathrooms', 'room_type'])
airbnb['price'] = airbnb['price'].str.replace('$', '')
airbnb['price'] = airbnb['price'].str.replace(',', '')
airbnb.price = pd.to_numeric(airbnb.price, errors='coerce')
airbnb = airbnb.dropna()

/tmp/ipykernel_171/722983309.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  airbnb['price'] = airbnb['price'].str.replace('$', '')


In [25]:
airbnb['bedrooms'].astype(float)
airbnb['bedrooms'] = np.where(airbnb.bedrooms >= 4, '4+', airbnb.bedrooms).astype(str)
airbnb['bedrooms'] = np.where(airbnb.bedrooms == 1, '1', airbnb.bedrooms).astype(str)
airbnb['bedrooms'] = np.where(airbnb.bedrooms == 2, '2', airbnb.bedrooms).astype(str)
airbnb['bedrooms'] = np.where(airbnb.bedrooms == 3, '3', airbnb.bedrooms).astype(str)

In [26]:
airbnb['log_price'] = np.log(airbnb.price + 1)

In [27]:
airbnb['accommodates'].astype(float)
airbnb['accommodates'] = np.where(airbnb.accommodates >= 4, '4 and more', airbnb.accommodates).astype(str)
airbnb['accommodates'] = np.where(airbnb.accommodates == 1, '1', airbnb.accommodates).astype(str)
airbnb['accommodates'] = np.where(airbnb.accommodates == 2, '2', airbnb.accommodates).astype(str)
airbnb['accommodates'] = np.where(airbnb.accommodates == 3, '3', airbnb.accommodates).astype(str)

In [28]:
airbnb.bathrooms = airbnb['bathrooms'].astype(float)
airbnb['rounded_bath'] = np.ceil(airbnb.bathrooms) #rounded all the .5 bathrooms up
airbnb['bathrooms_str'] = np.where(airbnb.rounded_bath == 0, '0', airbnb.rounded_bath).astype(str)
airbnb['bathrooms_str'] = np.where(airbnb.rounded_bath == 1, '1', airbnb.rounded_bath).astype(str)
airbnb['bathrooms_str'] = np.where(airbnb.rounded_bath == 2, '2', airbnb.rounded_bath).astype(str)
airbnb['bathrooms_str'] = np.where(airbnb.rounded_bath >= 3, '3 and more', airbnb.rounded_bath).astype(str)

## 2. (10pt) Now use the model above to predict (log) price for each listing in your data.

In [29]:
x = airbnb.drop(columns=['price', 'log_price', 'rounded_bath', 'bathrooms'])
y = airbnb.log_price

train_X, test_X, train_y, test_y = train_test_split(x,y, test_size=0.2)

m = smf.ols('log_price ~ bedrooms + accommodates + bathrooms_str + room_type', data=airbnb).fit()
predictions = m.predict(test_X)
predictions

34094    5.646500
35912    5.646500
1885     6.002294
38531    5.968708
25300    6.716133
           ...   
8129     5.646500
34326    6.280439
10132    5.968708
23761    6.716133
31511    5.646500
Length: 7739, dtype: float64

## 3. (10pt) Compute root-mean-squared-error (RMSE) of this prediction. RMSE is explained in lecture notes, 4.1.5 “Model evaluation: MSE, RMSE, R2”.

In [30]:
from sklearn.metrics import mean_squared_error
np.sqrt(mean_squared_error(test_y, predictions))

0.601819061529174

## 4. (10pt) Now use your model to predict the price for a 2-bedroom apartment that accommodates 4 (i.e. a full 2BR apartment).

In [31]:
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              log_price   R-squared:                       0.453
Model:                            OLS   Adj. R-squared:                  0.453
Method:                 Least Squares   F-statistic:                     2669.
Date:                Fri, 04 Mar 2022   Prob (F-statistic):               0.00
Time:                        01:02:42   Log-Likelihood:                -36034.
No. Observations:               38695   AIC:                         7.209e+04
Df Residuals:                   38682   BIC:                         7.220e+04
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       5.5593      0.062     89.221      0.000       5.437       5.681
bedrooms[T.1.0]                 0.0778      0.040      1.935      0.053      -0.001       0.157
bedrooms[T.2.0]                 0.2104      0.041      5.107      0.000       0.130       0.291
bedrooms[T.3.0]                 0.5135      0.043     12.062      0.000       0.430       0.597
bedrooms[T.4+]                  0.8933      0.044     20.263      0.000       0.807       0.980
accommodates[T.2]               0.3304      0.014     24.339      0.000       0.304       0.357
accommodates[T.3]               0.3942      0.017     23.512      0.000       0.361       0.427
accommodates[T.4 and more]      0.6085      0.016     39.040      0.000       0.578       0.639
bathrooms_str[T.1.0]            0.0013      0.047      0.028      0.978      -0.091       0.094
bathrooms_str[T.2.0]            0.0349      0.047      0.735      0.462      -0.058       0.128
bathrooms_str[T.3 and more]     0.5902      0.050     11.858      0.000       0.493       0.688
room_type[T.Private room]      -0.3222      0.007    -43.246      0.000      -0.337      -0.308
room_type[T.Shared room]       -0.9463      0.017    -55.789      0.000      -0.980      -0.913
==============================================================================
Omnibus:                     8589.601   Durbin-Watson:                   1.781
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            77599.711
Skew:                           0.814   Prob(JB):                         0.00
Kurtosis:                       9.744   Cond. No.                         52.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [32]:
5.5593 + 0.2104*1 + 0.6085

6.3782000000000005

## 5. (5pt) Compute the average log price for all listings in this group (2BR apartment that accommodates 4). Compare the result with your prediction. How close did you get?

In [33]:
a = airbnb[airbnb.accommodates == '4 and more']
a = airbnb[airbnb.bedrooms == '2.0']
np.mean(a.log_price)

6.343438554517532

I got a pretty close answer to my prediction from problem 2.4.